rclcpp/src/rclcpp/timer.cpp

```diff
@@ -138,3 +143,72 @@ TimerBase::exchange_in_use_by_wait_set_state(bool in_use_state)
 {
   return in_use_by_wait_set_.exchange(in_use_state);
 }
+
+void
+TimerBase::set_on_reset_callback(std::function<void()> callback)
+{
+  if (!callback) {
+    throw std::invalid_argument(
+            "The callback passed to set_on_reset_callback "
+            "is not callable.");
+  }
+
+  auto new_callback =
+    [callback, this](size_t reset_calls) {
+      try {
+        (void)reset_calls;
+        callback();
+      } catch (const std::exception & exception) {
+        RCLCPP_ERROR_STREAM(
+          rclcpp::get_logger("rclcpp"),
+          "rclcpp::TimerBase@" << this <<
+            " caught " << rmw::impl::cpp::demangle(exception) <<
+            " exception in user-provided callback for the 'on reset' callback: " <<
+            exception.what());
+      } catch (...) {
+        RCLCPP_ERROR_STREAM(
+          rclcpp::get_logger("rclcpp"),
+          "rclcpp::TimerBase@" << this <<
+            " caught unhandled exception in user-provided callback " <<
+            "for the 'on reset' callback");
+      }
+    };
+
+  std::lock_guard<std::recursive_mutex> lock(callback_mutex_);
+
+  // Set it temporarily to the new callback, while we replace the old one.
+  // This two-step setting, prevents a gap where the old std::function has
```

Comment:
why we cannot accept this gap with `callback_mutex_` ? what problem for user application could happen?

Reply:
This two-step setting is needed since `rcl` holds a pointer to the `rclcpp` callback `on_reset_callback_`.

If before assigning the `rcl` callback we assign the `rclcpp` one `on_reset_callback_ = new_callback`, then the `rcl` callback pointer would point (briefly) to a non-existing callback (segfault if called).
So we must first assign the `rcl` one, and then the `rclcpp` one (and since it has changed, we need again to set the `rcl` one back again).
Let me know if I wasn't clear enough on this.

Reply:
I see that. thank you so much for the explanation.